
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Network Analysis &#8212; tb-gene-signature-pipeline 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Machine Learning Pipeline" href="machine_learning_pipeline.html" />
    <link rel="prev" title="Welcome to tb-gene-signature-pipeline’s documentation!" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="network-analysis">
<h1>Network Analysis<a class="headerlink" href="#network-analysis" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="tb_gene_signature_pipeline.network_analysis.run_differential_expression_analysis">
<code class="sig-name descname">run_differential_expression_analysis</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">overwrite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">log_transform_all_geo_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tb_gene_signature_pipeline.network_analysis.run_differential_expression_analysis" title="Permalink to this definition">¶</a></dt>
<dd><p>Run marginal, per-gene differential expression analyses.</p>
<p>This function calls the differential analysis script
<code class="docutils literal notranslate"><span class="pre">tb-gene-signature-pipeline/R/differential_expression_analysis.R</span></code>.</p>
<p>An independent case/control differential expression analysis is run for
each dataset defined in <code class="docutils literal notranslate"><span class="pre">data/datasets.json</span></code>, and for each group comparison
defined in <code class="docutils literal notranslate"><span class="pre">data/comparisons.json</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><p>A dictionary with dataframes <cite>exprs</cite> and <cite>results</cite>.</p>
<p><cite>exprs</cite> contains preprocessed and normalized expression values
for each sample and across datasets.</p>
<p><cite>results</cite> contains differential expression effect sizes and p-values
for each dataset and case/control comparison.</p>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>dict of <code class="xref py py-class docutils literal notranslate"><span class="pre">pd.DataFrame</span></code></p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>For each GSE microarray experiment, the gene expression data pulled from GEO
is <span class="math notranslate nohighlight">\(\log_2\)</span> transformed, if the GEO data is not already <span class="math notranslate nohighlight">\(\log\)</span>
transformed. The transformed data is then quantile-normalized
within phenotype group* <a class="footnote-reference brackets" href="#id2" id="id1">1</a>.</p>
<p class="rubric">References</p>
<dl class="footnote brackets">
<dt class="label" id="id2"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Zhao, Y., Wong, L. &amp; Goh, W.W.B. How to do quantile normalization correctly for gene
expression data analyses. Sci Rep 10, 15534 (2020).
<a class="reference external" href="https://doi.org/10.1038/s41598-020-72664-6">https://doi.org/10.1038/s41598-020-72664-6</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tb_gene_signature_pipeline</span> <span class="k">as</span> <span class="nn">tb</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">diff_expr_results</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">run_differential_expression_analysis</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">diff_expr_results</span><span class="p">[</span><span class="s1">&#39;exprs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">         gene_symbol      gsm_id  preprocessed  normalized phenotype    dataset platform</span>
<span class="go">0               A1BG  GSM2712676        71.001    0.738524       atb  GSE101705   rnaseq</span>
<span class="go">1               A1BG  GSM2712677        65.999    0.947318       atb  GSE101705   rnaseq</span>
<span class="go">2               A1BG  GSM2712678        43.000    0.269351       atb  GSE101705   rnaseq</span>
<span class="go">3               A1BG  GSM2712679        41.000    0.310476       atb  GSE101705   rnaseq</span>
<span class="go">4               A1BG  GSM2712680        60.000    0.830032       atb  GSE101705   rnaseq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">diff_expr_results</span><span class="p">[</span><span class="s1">&#39;results&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">    dataset control case gene_symbol    log_fc  adj_p_val</span>
<span class="go">0  GSE19439      hc  atb        A1BG  0.026045   0.917217</span>
<span class="go">1  GSE19439      hc  atb        A1CF  0.091435   0.468395</span>
<span class="go">2  GSE19439      hc  atb         A2M  0.007356   0.964635</span>
<span class="go">3  GSE19439      hc  atb       A2ML1 -0.171059   0.204401</span>
<span class="go">4  GSE19439      hc  atb     A3GALT2 -0.008727   0.967318</span>
</pre></div>
</div>
</dd></dl>

<div class="line-block">
<div class="line"><br /></div>
</div>
<dl class="py function">
<dt id="tb_gene_signature_pipeline.network_analysis.merge_differential_expression_results">
<code class="sig-name descname">merge_differential_expression_results</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">differential_expression_df</span></em>, <em class="sig-param"><span class="n">adj_pval_thresh</span><span class="o">=</span><span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">log_fc_thresh</span><span class="o">=</span><span class="default_value">0.5849625007211562</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tb_gene_signature_pipeline.network_analysis.merge_differential_expression_results" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a <code class="xref py py-class docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> of differential expression results
(as generated by <a class="reference internal" href="#tb_gene_signature_pipeline.network_analysis.run_differential_expression_analysis" title="tb_gene_signature_pipeline.network_analysis.run_differential_expression_analysis"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_differential_expression_analysis()</span></code></a>), return
significant log fold change results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>differential_expression_df</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>) – A dataframe of differential expression results, as generated by
<a class="reference internal" href="#tb_gene_signature_pipeline.network_analysis.run_differential_expression_analysis" title="tb_gene_signature_pipeline.network_analysis.run_differential_expression_analysis"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_differential_expression_analysis()</span></code></a>.</p></li>
<li><p><strong>pval_thresh</strong> (<em>float</em>) – Adjusted p-value threshold for inclusion in merged dataframe.</p></li>
<li><p><strong>log_fc_thresh</strong> (<em>float</em>) – Effect size threshold for inclusion in merged dataframe.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>A <code class="docutils literal notranslate"><span class="pre">gene</span></code>-by-<code class="docutils literal notranslate"><span class="pre">dataset</span></code> dataframe of log fold change effect sizes
with the columns:</p>
<blockquote>
<div><ul class="simple">
<li><p>Name: control, dtype: object</p></li>
<li><p>Name: case, dtype: object</p></li>
<li><p>Name: gene_symbol, dtype: object</p></li>
<li><p>Name: <code class="docutils literal notranslate"><span class="pre">&lt;dataset</span> <span class="pre">GSE</span> <span class="pre">ID&gt;</span></code>, dtype: float (one column per dataset)</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">pd.DataFrame</span></code></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tb_gene_signature_pipeline</span> <span class="k">as</span> <span class="nn">tb</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_results</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">run_differential_expression_analysis</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">merged_results</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">merge_differential_expression_results</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adj_pval_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">log_fc_thresh</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="mf">1.5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">merged_results</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">  control case gene_symbol  GSE107994  GSE29536  GSE34608  GSE42825  ...</span>
<span class="go">0      hc  atb   1060P11.3        0.0       0.0  0.000000  0.000000  ...</span>
<span class="go">1      hc  atb     A2M-AS1        0.0       0.0  0.000000  0.000000  ...</span>
<span class="go">2      hc  atb       AAED1        0.0       0.0  0.862417  0.000000  ...</span>
<span class="go">3      hc  atb       AAMDC        0.0       0.0  0.000000  0.694117  ...</span>
<span class="go">4      hc  atb        AAMP        0.0       0.0  0.000000  0.000000  ...</span>
</pre></div>
</div>
</dd></dl>

<div class="line-block">
<div class="line"><br /></div>
</div>
<dl class="py function">
<dt id="tb_gene_signature_pipeline.network_analysis.compute_edge_weight_distributions">
<code class="sig-name descname">compute_edge_weight_distributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">merged_results</span></em>, <em class="sig-param"><span class="n">overwrite</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tb_gene_signature_pipeline.network_analysis.compute_edge_weight_distributions" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<div class="line-block">
<div class="line"><br /></div>
</div>
<dl class="py function">
<dt id="tb_gene_signature_pipeline.network_analysis.generate_null_edge_weight_distributions">
<code class="sig-name descname">generate_null_edge_weight_distributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">merged_results</span></em>, <em class="sig-param"><span class="n">n_iter</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">overwrite</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tb_gene_signature_pipeline.network_analysis.generate_null_edge_weight_distributions" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<div class="line-block">
<div class="line"><br /></div>
</div>
<dl class="py function">
<dt id="tb_gene_signature_pipeline.network_analysis.construct_networks">
<code class="sig-name descname">construct_networks</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">merged_results</span></em>, <em class="sig-param"><span class="n">overwrite</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tb_gene_signature_pipeline.network_analysis.construct_networks" title="Permalink to this definition">¶</a></dt>
<dd><p>Construct networks based on shared association signal across
differential expression analysis datasets.</p>
<p>One network is created for each case/control comparison
(<code class="docutils literal notranslate"><span class="pre">hc</span></code> vs. <code class="docutils literal notranslate"><span class="pre">atb</span></code>, etc.).</p>
<p>Within each network, a node represents a gene, and an edge between
nodes represents indicates that those two genes had significant
differential expression associations in the <strong>same</strong> direction in
<strong>at least 3</strong> datasets.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>merged_results</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>) – A gene-by-dataset dataframe of significant log fold change
effect sizes, as produced by <a class="reference internal" href="#tb_gene_signature_pipeline.network_analysis.merge_differential_expression_results" title="tb_gene_signature_pipeline.network_analysis.merge_differential_expression_results"><code class="xref py py-func docutils literal notranslate"><span class="pre">merge_differential_expression_results()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl>
<dt>A dictionary with the entries:</dt><dd><dl>
<dt>graphs<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">pd.DataFrame</span></code></span></dt><dd><dl class="simple">
<dt>Columns:</dt><dd><ul class="simple">
<li><p>Name: control, dtype: object</p></li>
<li><p>Name: case, dtype: object</p></li>
<li><p>Name: graph, dtype: <code class="xref py py-class docutils literal notranslate"><span class="pre">nx.Graph</span></code></p></li>
</ul>
</dd>
</dl>
</dd>
<dt>nodes<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">pd.DataFrame</span></code></span></dt><dd><dl class="simple">
<dt>Columns:</dt><dd><ul class="simple">
<li><p>Name: control, dtype: object</p></li>
<li><p>Name: case, dtype: object</p></li>
<li><p>Name: gene_symbol, dtype: object</p></li>
<li><p>Name: degree, dtype: float</p></li>
<li><p>Name: weighted_degree, dtype: float</p></li>
<li><p>Name: eigenvector_centrality, dtype: float</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tb_gene_signature_pipeline</span> <span class="k">as</span> <span class="nn">tb</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">differential_expression_results</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">run_differential_expression_analysis</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">merged_results</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">merge_differential_expression_results</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">differential_expression_results</span><span class="p">,</span> <span class="n">adj_val_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">log_fc_thresh</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="mf">1.5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">networks</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">construct_networks</span><span class="p">(</span><span class="n">merged_results</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">networks</span><span class="p">[</span><span class="s1">&#39;graphs&#39;</span><span class="p">])</span>
<span class="go">  control  case                                              graph</span>
<span class="go">0      hc   atb  (AAMDC, ABCA1, ACOT8, ACOT9, ACSL1, ACTA2, ADA...</span>
<span class="go">1      hc  ltbi                                     (ATP1B2, ETV7)</span>
<span class="go">2    ltbi   atb  (ABCA1, ABCA13, ABCC13, ACSL1, ACSL4, ADAM9, A...</span>
<span class="go">3      od   atb  (ADM, AIM2, ANKRD22, APOL6, ATF3, BATF2, BRSK1...</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">networks</span><span class="p">[</span><span class="s1">&#39;nodes&#39;</span><span class="p">])</span>
<span class="go">     control case gene_symbol  degree  weighted_degree  eigenvector_centrality</span>
<span class="go">0         hc  atb       AAMDC     503        94.312500            1.404467e-02</span>
<span class="go">1         hc  atb       ABCA1    1318       349.187500            4.088542e-02</span>
<span class="go">2         hc  atb       ACOT8     503        94.312500            1.404467e-02</span>
<span class="go">3         hc  atb       ACOT9    1099       271.000000            3.351424e-02</span>
<span class="go">4         hc  atb       ACSL1    1327       354.187500            4.143864e-02</span>
<span class="gp">... </span>     <span class="o">...</span>  <span class="o">...</span>         <span class="o">...</span>     <span class="o">...</span>              <span class="o">...</span>                     <span class="o">...</span>
<span class="go">3409      od  atb        RGL1       5         1.666667            1.456176e-03</span>
<span class="go">3410      od  atb        TLR5       5         1.666667            1.456176e-03</span>
<span class="go">3411      od  atb        CD3G       4         1.333333            1.917126e-11</span>
<span class="go">3412      od  atb        GNLY       4         1.333333            1.917126e-11</span>
<span class="go">3413      od  atb        NRG1       4         1.333333            1.917126e-11</span>
</pre></div>
</div>
</dd></dl>

<div class="line-block">
<div class="line"><br /></div>
</div>
<dl class="py function">
<dt id="tb_gene_signature_pipeline.network_analysis.generate_gene_lists">
<code class="sig-name descname">generate_gene_lists</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nodes</span></em>, <em class="sig-param"><span class="n">top_n</span><span class="o">=</span><span class="default_value">100</span></em><span class="sig-paren">)</span><a class="headerlink" href="#tb_gene_signature_pipeline.network_analysis.generate_gene_lists" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate gene lists for each comparison by returning the
top nodes in each comparison network by <code class="docutils literal notranslate"><span class="pre">weighted_degree</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nodes</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">pd.DataFrame</span></code>) – A dataframe of networks node metrics, as returned by the <code class="docutils literal notranslate"><span class="pre">nodes</span></code>
entry of <code class="xref py py-func docutils literal notranslate"><span class="pre">generate_networks()</span></code>.</p></li>
<li><p><strong>top_n</strong> (<em>int</em>) – Return the <code class="docutils literal notranslate"><span class="pre">top_n</span></code> nodes in each comparison network.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A dictionary of gene lists, keyed by comparison (e.g. <code class="docutils literal notranslate"><span class="pre">'hc-atb'</span></code>.)</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">tb_gene_signature_pipeline.network_analysis</span> <span class="k">as</span> <span class="nn">tb</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">diff_expr_results</span> <span class="o">=</span> <span class="n">na</span><span class="o">.</span><span class="n">run_differential_expression_analysis</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">log_transform_all_geo_data</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">merged_results</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">merge_differential_expression_results</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">differential_expression_results</span><span class="p">,</span> <span class="n">adj_val_thresh</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">log_fc_thresh</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="mf">1.5</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">networks</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">construct_networks</span><span class="p">(</span><span class="n">merged_results</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gene_lists</span> <span class="o">=</span> <span class="n">tb</span><span class="o">.</span><span class="n">generate_gene_lists</span><span class="p">(</span><span class="n">networks</span><span class="p">[</span><span class="s1">&#39;nodes&#39;</span><span class="p">],</span> <span class="n">top_n</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<div class="line-block">
<div class="line"><br /></div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">tb-gene-signature-pipeline</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Network Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="machine_learning_pipeline.html">Machine Learning Pipeline</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Welcome to tb-gene-signature-pipeline’s documentation!</a></li>
      <li>Next: <a href="machine_learning_pipeline.html" title="next chapter">Machine Learning Pipeline</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Liam Abbott.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.2.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/network_analysis.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>